example/particles/particles.cpp

```diff
@@ -316,39 +318,176 @@ TaskStatus Defrag(MeshBlock *pmb) {
+  status = ftc.Execute();
+
+  return status;
+}
+
+// TODO(BRR) This should really be in parthenon/src... but it can't just live in Swarm
+// because of the loop over blocks
+TaskStatus StopCommunicationMesh(BlockList_t &blocks) {
+  int num_sent_local = 0;
+  for (auto &block : blocks) {
+    auto &pmb = block;
+    auto sc = pmb->swarm_data.Get();
+    auto swarm = sc->Get("my particles");
+    swarm->finished_transport = false;
+    num_sent_local += swarm->num_particles_sent_;
+  }
+
+  // Boundary transfers on same MPI proc are blocking
+  for (auto &block : blocks) {
+    auto swarm = block->swarm_data.Get()->Get("my particles");
+    for (int n = 0; n < block->pbval->nneighbor; n++) {
+      NeighborBlock &nb = block->pbval->neighbor[n];
+      // TODO(BRR) May want logic like this if we have non-blocking TaskRegions
+      // if (nb.snb.rank != Globals::my_rank) {
+      //  if (swarm->vbswarm->bd_var_.flag[nb.bufid] != BoundaryStatus::completed) {
+      //    printf("[%i] Neighbor %i not complete!\n", Globals::my_rank, n);
+      //    //return TaskStatus::incomplete;
+      //  }
+      //}
+
+      // TODO(BRR) May want to move this logic into a per-cycle initialization call
+      if (swarm->vbswarm->bd_var_.flag[nb.bufid] == BoundaryStatus::completed) {
+        swarm->vbswarm->bd_var_.req_send[nb.bufid] = MPI_REQUEST_NULL;
+      }
+    }
+  }
+
+  int num_sent_global = 0;
+  MPI_Allreduce(&num_sent_local, &num_sent_global, 1, MPI_INT, MPI_SUM, MPI_COMM_WORLD);
+
+  if (num_sent_global == 0) {
+    for (auto &block : blocks) {
+      auto &pmb = block;
+      auto sc = pmb->swarm_data.Get();
+      auto swarm = sc->Get("my particles");
+      swarm->finished_transport = true;
+
+      // TODO(BRR) should this really be done at an initialization step for each cycle?
+      for (int n = 0; n < swarm->vbswarm->bd_var_.nbmax; n++) {
+        auto &nb = pmb->pbval->neighbor[n];
+        swarm->vbswarm->bd_var_.flag[nb.bufid] = BoundaryStatus::waiting;
+      }
+    }
+  }
+
+  for (auto &block : blocks) {
+    auto &pmb = block;
+    auto sc = pmb->swarm_data.Get();
+    auto swarm = sc->Get("my particles");
+    swarm->finished_transport = true;
+    for (int n = 0; n < swarm->vbswarm->bd_var_.nbmax; n++) {
+      auto &nb = pmb->pbval->neighbor[n];
+      swarm->vbswarm->bd_var_.flag[nb.bufid] = BoundaryStatus::waiting;
+    }
+  }
+
+  return TaskStatus::complete;
+}
+
+TaskCollection ParticleDriver::MakeParticlesCreationTaskCollection() {
   TaskCollection tc;
+  TaskID none(0);
+  double t0 = tm.time;
+  BlockList_t &blocks = pmesh->block_list;
+
+  auto num_task_lists_executed_independently = blocks.size();
+  TaskRegion &async_region0 = tc.AddRegion(num_task_lists_executed_independently);
+  for (int i = 0; i < blocks.size(); i++) {
+    auto &pmb = blocks[i];
+    auto &tl = async_region0[i];
+    auto create_some_particles = tl.AddTask(none, CreateSomeParticles, pmb.get(), t0);
+  }
```

Comment:
I worry you might risk some flexibility in doing that -- in `examples/advection/advection_driver.cpp`, for example, you do this but also add containers in this loop over blocks. Also, at least right now I like how explicit this is, since I already seem to take wrong turns pretty easily when thinking about making task lists. What you say could still be a win, I'm just typing out what's in my brain now here.

Reply:
yeah ok. I think the templated pattern is a broader discussion for the tasking infrastructure. Doesn't have to be addressed here.